src/oic/utils/sdb.py

```diff
@@ -148,11 +156,23 @@ def is_expired(self, token, when=None):
         return bool(now > eat)
 
     def invalidate(self, token):
-        pass
+        """Mark the refresh token as invalidated."""
+        if self.get_type(token) != "R":
+            return
+        sid = self.get_key(token)
+        self.token_storage[sid]["revoked"] = True
 
     def valid(self, token):
-        self.type_and_key(token)
-        return True
+        try:
+            typ, key = self.type_and_key(token)
+        except (Error, InvalidToken):
+            raise WrongTokenType()
+        if typ != self.type:
+            raise WrongTokenType()
+        if typ == "R":
+            return not self.token_storage[key].get("revoked", False)
```

Comment:
Should this care for KeyErrors, e.g. token has a key, but key isn't present in the token_storage?

Reply:
I was thinking about the same. My original though was to have an ABC with an API and handle one specific error, but we already have several levels of ABCs...

So unless we want to add another layer of ABC for API, I think keeping this is probably better.

Reply:
Agreed.